platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java

```diff
@@ -225,15 +232,21 @@ private static CodeBlock getGetCall(String propertyGetterName, CallableReturnTyp
             case FILE_COLLECTION:
                 return CodeBlock.of("return $N.$N()", SELF_PARAMETER_NAME, propertyGetterName);
             case LIST_PROPERTY:
-                return CodeBlock.of("return new $T<>($N.$N())", LIST_PROPERTY_LIST_VIEW.asTypeName(), SELF_PARAMETER_NAME, propertyGetterName);
+                return CodeBlock.of("return new $T<>($N.$N())", LIST_PROPERTY_LIST_VIEW.asClassName(), SELF_PARAMETER_NAME, propertyGetterName);
             case SET_PROPERTY:
-                return CodeBlock.of("return new $T<>($N.$N())", SET_PROPERTY_SET_VIEW.asTypeName(), SELF_PARAMETER_NAME, propertyGetterName);
+                return CodeBlock.of("return new $T<>($N.$N())", SET_PROPERTY_SET_VIEW.asClassName(), SELF_PARAMETER_NAME, propertyGetterName);
             case MAP_PROPERTY:
-                return CodeBlock.of("return new $T<>($N.$N())", MAP_PROPERTY_MAP_VIEW.asTypeName(), SELF_PARAMETER_NAME, propertyGetterName);
+                return CodeBlock.of("return new $T<>($N.$N())", MAP_PROPERTY_MAP_VIEW.asClassName(), SELF_PARAMETER_NAME, propertyGetterName);
             case PROPERTY:
                 return CodeBlock.of("return $N.$N().getOrElse($L)", SELF_PARAMETER_NAME, propertyGetterName, TypeUtils.getDefaultValue(returnType.getType()));
+            case PROVIDER: {
+                TypeName newPropertyType = implementationExtra.getNewPropertyType();
+                return newPropertyType instanceof ParameterizedTypeName && isAssignableToFileSystemLocation(((ParameterizedTypeName) newPropertyType).typeArguments.get(0))
```

Comment:
💅 Let's name this type parameter to make this line more readable. Can we use a utility method to get it?

Reply:
Added `TypeUtils. getTypeParameter(TypeName typeName, int index)` that return `Optional<TypeName>`, it's a bit nicer now I think